libbeat/publisher/pipeline/client_worker.go

```diff
@@ -115,6 +115,9 @@ func (w *netClientWorker) run() {
 		reconnectAttempts = 0
 	)
 
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
It looks as if we could replace the `done chan struct{}` channel in the worker struct with a context's cancel func, by moving the context creation into `makeClientWorker`.

This would make it so that calling `Close()` aborts in progress network calls immediately. Maybe this is a negative though, the way it is now we finish any in progress work before exiting but it isn't clear how long we'd wait for it.

The semantics of the done channel here mean once `Close()` is called we'll never publish a batch again after the current batch succeeds so this feels like it should be safe to do if we accept aborting an in progress batch as a consequence.

Reply:
I think you might have to make this change if the windows shutdown bug was caused by the network request taking too long to complete. "Abort connection on close" was the logic of the original change for that fix.

Reply:
> It looks as if we could replace the done chan struct{} channel in the worker struct with a context's cancel func, by moving the context creation into makeClientWorker.

It makes sense, the current implementation already cancel all in-flight requests when `Close` is called, because it causes `run` to return, cancelling the context from the connection.

I'll make this change, I don't see any change in behaviour from the current state of this PR.

Thanks!

Reply:
Can you write a test that proves that calling `close()` causes long running network requests to abort? You can probably use mockES to just hold an _bulk open for a ridiculously long time. 

I think that is the source of the problem in https://github.com/elastic/beats/issues/40518, and that we were able to write quite a lot of code without quite fixing it means we should add a test to prevent regressing on this.